elfpy/bots/checkpoint_bot.py

```diff
@@ -0,0 +1,125 @@
+"""A checkpoint bot for Hyperdrive"""
+from __future__ import annotations
+
+import datetime
+from dotenv import load_dotenv
+import logging
+import os
+import time
+
+from web3.contract.contract import Contract
+
+from elfpy import eth, hyperdrive_interface
+from elfpy.agents.agent import Agent
+from elfpy.eth.accounts.eth_account import EthAccount
+from elfpy.eth.rpc_interface import set_anvil_account_balance
+from elfpy.eth.transactions import smart_contract_read, smart_contract_transact
+from elfpy.utils import logs
+from elfpy.bots.environment_config import EnvironmentConfig
+from elfpy.hyperdrive_interface.hyperdrive_interface import get_hyperdrive_config
+
+# The portion of the checkpoint that the bot will wait before attempting to
+# mint a new checkpoint.
+CHECKPOINT_WAITING_PERIOD = 0.5
+
+
+def get_config() -> EnvironmentConfig:
+    # Load some configuration variables from the environment.
+    load_dotenv()
+    artifacts_url = os.environ.get("ARTIFACTS_URL")
+    if artifacts_url is None:
+        raise ValueError("ARTIFACTS_URL environment variable must be set")
+    rpc_url = os.environ.get("RPC_URL")
+    if rpc_url is None:
+        raise ValueError("RPC_URL environment variable must be set")
+
+    # The configuration for the checkpoint bot halts on errors and logs to stdout.
+    return EnvironmentConfig(
+        # Networking
+        artifacts_url=artifacts_url,
+        rpc_url=rpc_url,
+        # Errors
+        halt_on_errors=True,
+        # Logging
+        log_stdout=True,
+        log_level=logging.INFO,
+    )
+
+
+def does_checkpoint_exist(hyperdrive: Contract, checkpointTime: int) -> bool:
+    return smart_contract_read(hyperdrive, "getCheckpoint", checkpointTime)["sharePrice"] > 0
+
+
+def main() -> None:
+    # Get the configuration and initialize the web3 provider.
+    config = get_config()
+    provider = eth.web3_setup.initialize_web3_with_http_provider(config.rpc_url, reset_provider=False)
```

Comment:
We've been calling this `web3` elsewhere. The web3 docs seem to oscillate randomly between `w3` and `web3`, but we like the latter because it conforms to pep8 linting standards. The "provider" is a communication component (how web3 talks to the chain). The thing returned is a `Web3` object, which is not given a name afaik, but I guess could be called a "container" or "wrapper". More here: https://web3py.readthedocs.io/en/latest/internals.html#web3-internals

Reply:
Ok, I'm fine with moving back to `web3`. I called it `provider` since we use it like a provider (this is how we make RPC requests and interact with the ethereum node), but I see your point.